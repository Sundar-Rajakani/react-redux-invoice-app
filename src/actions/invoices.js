import uuid from 'uuid';

export const addInvoice = ({ description = '', note = '', amount = 0, createdAt = 0 } = {}) => ({
  type: 'ADD_INVOICE',
  invoice: {
    id: uuid(),
    description,
    note,
    createdAt,
    amount
  }
})

// export const editInvoice = (id, updates) => ({
//   type: 'EDIT_INVOICE',
//   id,
//   updates
// })


export const editInvoice = (id, updates) => {
  return {
    type: 'EDIT_INVOICE',
    id,
    updates
  }
}


export const removeInvoice = ({ id }) => ({
  type: 'REMOVE_INVOICE',
  id
});