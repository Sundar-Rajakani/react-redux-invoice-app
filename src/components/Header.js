import React from 'react';
import { NavLink } from 'react-router-dom';
const Header = () => (
  <header>
    <h1>Invoice Management</h1>
    <NavLink to="/" activeClassName="isActive" exact={true}>Home</NavLink>
    <NavLink to="/add" activeClassName="isActive">AddInvoice</NavLink>
    <NavLink to="/help" activeClassName="isActive">Help</NavLink>
  </header>
);

export default Header;
