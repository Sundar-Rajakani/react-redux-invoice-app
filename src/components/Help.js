import React from 'react';


const Help = () => (
  <div>This is help hage</div>
);

export default Help;
