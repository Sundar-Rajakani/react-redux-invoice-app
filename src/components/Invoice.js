import React from 'react';
import { Link } from 'react-router-dom';

export const Invoice = ({ description, amount, createdAt, id, dispatch }) => (
  <div>
    <Link to={`/edit/${id}`}>
      <h3>{description}</h3>
    </Link>
    <p>{amount} - {createdAt}</p>

  </div>
);

export default Invoice;
