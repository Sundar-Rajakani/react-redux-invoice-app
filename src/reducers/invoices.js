const invoicesReducerDefaultState = [];

const InvoicesReducer = (state = [], action) => {
  switch (action.type) {
    case 'ADD_INVOICE':
      return [...state, action.invoice];
    case 'EDIT_INVOICE':
      return state.map((invoice) => {
        if (invoice.id === action.id) {
          return {
            ...invoice,
            ...action.updates
          }
        }
        else {
          return invoice;
        }
      });
    case 'REMOVE_INVOICE':
      return state.filter(({ id }) => id !== action.id);
    default:
      return state;
  }
}

export default InvoicesReducer;