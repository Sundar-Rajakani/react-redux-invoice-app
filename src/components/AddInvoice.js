import React, { Component } from 'react';
import { connect } from 'react-redux';

import { addInvoice } from '../actions/invoices'
import InvoiceForm from './InvoiceForm';


export class AddInvoice extends Component {

  onSubmit = (invoice) => {
    this.props.addInvoice(invoice);  
    this.props.history.push('/');
  }

  render() {
    return (
      <div>
        <h1>Invoice Form</h1>
        <InvoiceForm
          onSubmit={this.onSubmit} />
      </div>
    )
  }
}


const mapDispatchToProps = (dispatch) => ({
  addInvoice: (invoice) => dispatch.addInvoice(invoice)
})
export default connect(undefined, mapDispatchToProps)(AddInvoice);
