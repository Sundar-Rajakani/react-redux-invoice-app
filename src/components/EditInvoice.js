import React, { Component } from 'react';
import { connect } from 'react-redux';

import { editInvoice, removeInvoice } from '../actions/invoices';
import InvoiceForm from './InvoiceForm';



export class EditInvoice extends Component {


  onSubmit = (invoice) => {
    this.props.editInvoice(props.invoice.id, invoice);
    this.props.history.push('/');
  }

  onClick = (e) => {
    this.props.removeInvoice({ id: props.invoice.id });
    this.props.history.push('/');
  }

  render() {
    return (
      <div>
        <h1>Edit Invoice</h1>
        <InvoiceForm
          invoice={this.props.invoice}
          onSubmit={this.onSubmit} />
        <button onClick={this.onClick}>Remove</button>
      </div >
    );
  }
}

const mapStateToProps = (state, props) => {
  return {
    invoice: state.invoices.find((invoice) => invoice.id === props.match.params.id)
  }
}

const mapDispatchToProps = (dispatch, props) => ({
  editInvoice: (id, invoice) => dispatch(editInvoice(id, invoice)),
  removeInvoice: (data) => dispatch(removeInvoice(data))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditInvoice);
