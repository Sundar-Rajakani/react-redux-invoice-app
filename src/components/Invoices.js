import React from 'react';
import { connect } from 'react-redux';

import getFilteredInvoices from '../selectors/invoices';
import Invoice from './Invoice';

export const Invoices = (props) => (
  <div>
    <h1>Invoices </h1>
    {
      props.invoices.length === 0 ? (
        <p>No Invoices found</p>
      ) : (
        props.invoices.map((invoice) => {
          return <Invoice key={invoice.id} {...invoice} />
        })
      )
    }
  </div>
)

const mapStateToProps = (state) => {
  return {
    invoices: getFilteredInvoices(state.invoices, state.filters)
  }
}

export default connect(mapStateToProps)(Invoices);

