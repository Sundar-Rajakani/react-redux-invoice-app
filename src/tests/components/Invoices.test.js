import React from 'react';
import { shallow } from 'enzyme';
import { Invoices } from '../../components/Invoices';
import invoices from '../fixtures/invoices';

test('should render list of invoices with the given data', () => {
  const wrapper = shallow(<Invoices invoices={invoices}/>);
  expect(wrapper).toMatchSnapshot();
});

test('should display a message when there are not invoices', () => {
  const wrapper = shallow(<Invoices invoices={[]}/>);
  expect(wrapper).toMatchSnapshot();
});
