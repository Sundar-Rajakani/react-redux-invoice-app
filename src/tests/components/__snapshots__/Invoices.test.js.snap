// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should display a message when there are not invoices 1`] = `
<div>
  <h1>
    Invoices 
  </h1>
  <p>
    No Invoices found
  </p>
</div>
`;

exports[`should render list of invoices with the given data 1`] = `
<div>
  <h1>
    Invoices 
  </h1>
  <Invoice
    amount={4250}
    createdAt={0}
    description="desc1"
    id="1"
    key="1"
    note="note1"
  />
  <Invoice
    amount={2503}
    createdAt={-345600000}
    description="desc2"
    id="2"
    key="2"
    note="note2"
  />
  <Invoice
    amount={2500}
    createdAt={345600000}
    description="desc3"
    id="3"
    key="3"
    note="note3"
  />
</div>
`;
